### PR TITLE
Generate: documented function to compute the transition scores

### LINKED ISSUE #18616: BartForConditionalGeneration is erroneous either at .forward or at .generate
### System Info

- `transformers` version: 4.20.1
- Platform: Linux-5.4.0-58-generic-x86_64-with-debian-buster-sid
- Python version: 3.7.12
- Huggingface_hub version: 0.8.1
- PyTorch version (GPU?): 1.10.1+cu111 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: True
- Using distributed or parallel set-up in script?: False

### Who can help?

@patil-suraj @patrickvonplaten

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

```
from transformers import AutoModelForSeq2SeqLM, AutoTokenizer
import torch

text = """
Phillip,   Could you please do me a favor?\nI would like  to read your current title policy to see what \
it says about easements.\nYou  should have received a copy during your closing.\nI don't know how many \
pages it will be but let me know how you want to handle getting a copy  made.\nI'll be happy to make the copy,\
or whatever makes it easy for  you.\nThanks,\n
"""

checkpoint = "Aktsvigun/bart-base_aeslc_42"
model = AutoModelForSeq2SeqLM.from_pretrained(checkpoint).cuda()
tokenizer = AutoTokenizer.from_pretrained(checkpoint)

input_ids = tokenizer(text, truncation=True, return_tensors="pt")["input_ids"].to(model.device)

generate_output = model.generate(
    input_ids, num_return_sequences=4, length_penalty=1., return_dict_in_generate=True, output_scores=True, early_stopping=True
)

# Most probable labels according to the generate output. Taking from first since do not need initial generation token.
labels = generate_output.sequences[0][generate_output.sequences[0] != 1][None, 1:]
out = model(input_ids, labels=labels)
probas = torch.nn.functional.softmax(out.logits, dim=-1)

sequence_score = probas[0].log().gather(index=labels[0][:, None], dim=-1).sum() / len(labels[0])
assert torch.allclose(-sequence_score, out.loss)
assert torch.allclose(sequence_score, generate_output.sequences_scores[0])

```

### Expected behavior

The last assert must be passed, yet the results differ (-0.8670 for reconstructed score and -0.8581 from generated output). What happens in the code: I first generate the sequence with BART, and then I try to reproduce the score by calling `.forward` (reproducing the score as the average of log-probas of labels ids taken from each decoder iteration).

Why is it important: this is a "sub-bug" which I found, verifying another bug: I wrote a function to restore the sequences and sequences scores from `transformers.generation_utils.BeamSearchEncoderDecoderOutput.scores` and got slightly different results with the ones outputted by `transformers.generation_utils.BeamSearchEncoderDecoderOutput`. Namely, I restore some sequences with the scores, higher than `transformers.generation_utils.BeamSearchEncoderDecoderOutput.sequences_scores`. I need to check, which version (default / mine) is correct, hence I need to pass the sequence with forward and calculate its "intrinsic" score. However, as this example shows, either `.forward` or `.generate` return slightly erroneous results.

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

Fixes #18616; Addresses comments in #5164, #20008 (and in a few other issues that I've lost track of).

## The issue

A few users would like to have a simple function to obtain the transition scores (i.e. the logits for each selected token at generate time). This is very useful for exploring the generated contents and simplifies the construction of powerful color-coded interfaces (e.g. [this one](https://joel.tools/codegen/)). It is also commonly requested to compare our models against OpenAI's.

We had a function for that in PT, `compute_beam_transition_scores`, but it was unknown to most users. This is because it was limited to Beam-based approaches, was not in our documents, and had no examples.

## The solution

This PR upgrades the function above to a first-class citizen 🥇 :
1. Makes it compatible with all generation strategies (e.g. Sample)
2. Adds a flag to renormalize the logits before fetching the right ones, which is a frequent downstream use
3. Adds it to the documentation
4. Populates the docstring with examples for which I've got questions a few times (how to print the token probabilities and how to recompute the score of the sequences in beam search)

In the process, I've decided to update the name of the function (`compute_beam_transition_scores` -> `compute_transition_scores`), to match better what it does. Although this technically breaks the API, the function was not part of our documented functions and, given the number of related issues, I'd say it was mostly unknown.

### CODE REVIEW

File: src/transformers/generation/utils.py [hunk 1]
```diff
@@ -914,42 +914,121 @@ def _merge_criteria_processor_list(
...
+        ... )
+        >>> # If you sum the generated tokens' scores and apply the length penalty, you'll get the sequence scores.
+        >>> # Tip: set `normalize_logits=True` to recompute the scores from the normalized logits.
+        >>> output_length = inputs.input_ids.shape[1] + np.sum(transition_scores.numpy() < 0, axis=1)
+        >>> length_penalty = model.generation_config.length_penalty
+        >>> reconstructed_scores = transition_scores.sum(axis=1) / (output_length**length_penalty)
+        >>> print(np.allclose(outputs.sequences_scores, reconstructed_scores))
+        True
+        ```"""
+        # 1. In abcense of `beam_indices`, we can assume that we come from e.g. greedy search, which is equivalent
```
Comment (sgugger) [inline src/transformers/generation/utils.py:995]:
```suggestion
        # 1. In absence of `beam_indices`, we can assume that we come from e.g. greedy search, which is equivalent
```

File: tests/generation/test_utils.py [hunk 2]
```diff
@@ -2485,6 +2487,64 @@ def test_generate_encoder_outputs_attention_mask(self):
 
         self.assertListEqual(output_sequences_no_mask.tolist(), output_sequences_with_mask.tolist())
 
+    def test_transition_scores_greedy_search(self):
+        articles = [
+            "Justin Timberlake",
+            "Michael Phelps",
+        ]
```
Comment (sgugger) [inline tests/generation/test_utils.py:2494]: Fits in one line.

File: tests/generation/test_utils.py [hunk 3]
```diff
@@ -2485,6 +2487,64 @@ def test_generate_encoder_outputs_attention_mask(self):
...
+                [0.34934354, 0.4935004, 0.6373219, 0.5173545, 0.57517034],
+            ]
+        )
+        self.assertTrue(np.allclose(transition_scores.cpu().numpy(), expected_scores))
+
+    def test_transition_scores_greedy_search_normalized(self):
+        articles = [
+            "Justin Timberlake",
+            "Michael Phelps",
+        ]
```
Comment (sgugger) [inline tests/generation/test_utils.py:2523]: Same here.

**Review by sgugger — APPROVED** [review]
Thanks for working on this, just left a couple of nits.

File: src/transformers/generation/utils.py [hunk 4]
```diff
@@ -914,42 +914,121 @@ def _merge_criteria_processor_list(
...
 
-        # 1. reshape scores as [vocab_size * batch_size, # generation steps]
-        # with batch_size being 2 * vocab_size and # generation steps being
+        Return:
+            `torch.Tensor`: A `torch.Tensor` of shape `(batch_size*num_return_sequences, sequence_length)` containing
+                the transition scores (logits)
+
+        Examples:
+
+        ```python
```
Comment (patrickvonplaten) [inline src/transformers/generation/utils.py:950]: super nice example!

File: src/transformers/generation/utils.py [hunk 5]
```diff
@@ -914,42 +914,121 @@ def _merge_criteria_processor_list(
...
+        """
+        Computes the transition scores of sequences given the generation scores (and beam indices, if beam search was
+        used). This is a convenient method to quicky obtain the scores of the selected tokens at generation time.
+
+        Parameters:
+            sequences (`torch.LongTensor`):
+                The generated sequences. The second dimension (sequence_length) is either equal to `max_length` or
+                shorter if all batches finished early due to the `eos_token_id`.
+            scores (`tuple(torch.FloatTensor)`):
+                Beam transition scores for each vocabulary token at each generation step. Beam transition scores
```
Comment (patrickvonplaten) [inline src/transformers/generation/utils.py:933]:
```suggestion
                Transition scores for each vocabulary token at each generation step. Beam transition scores
```
The function should be used for both beam search and non-beam search outputs no?

File: src/transformers/generation/utils.py [hunk 6]
```diff
@@ -914,42 +914,121 @@ def _merge_criteria_processor_list(
...
+                consisting of log probabilities of tokens conditioned on log softmax of previously generated tokens in
+                this beam. Tuple of `torch.FloatTensor` with up to `max_new_tokens` elements (one element for each
```
Comment (patrickvonplaten) [inline src/transformers/generation/utils.py:935]:
```suggestion
                Tuple of `torch.FloatTensor` with up to `max_new_tokens` elements (one element for each
```

File: src/transformers/generation/utils.py [hunk 7]
```diff
@@ -914,42 +914,121 @@ def _merge_criteria_processor_list(
...
+                generated token), with each tensor of shape `(batch_size*num_beams, config.vocab_size)`.
+            beam_indices (`tuple(tuple(torch.LongTensor))`, *optional*):
+                Beam indices of generated token id at each generation step. `torch.LongTensor` of shape
+                `(batch_size*num_return_sequences, input_ids.shape[-1])`. Only required if a `num_beams>1` at
+                generate-time.
+            normalize_logits (`bool`, *optional*, defaults to `True`):
```
Comment (patrickvonplaten) [inline src/transformers/generation/utils.py:941]: In the code it defaults to `False` (maybe change here? )
```suggestion
            normalize_logits (`bool`, *optional*, defaults to `False`):
```

File: src/transformers/generation/utils.py [hunk 8]
```diff
@@ -914,42 +914,121 @@ def _merge_criteria_processor_list(
...
+        Examples:
+
+        ```python
+        >>> from transformers import GPT2Tokenizer, AutoModelForCausalLM
+        >>> import numpy as np
+
+        >>> tokenizer = GPT2Tokenizer.from_pretrained("gpt2", padding_side="left")
+        >>> model = AutoModelForCausalLM.from_pretrained("gpt2")
+        >>> tokenizer.pad_token_id = tokenizer.eos_token_id
+        >>> inputs = tokenizer(["Today is"], return_tensors="pt", padding=True)
```
Comment (patrickvonplaten) [inline src/transformers/generation/utils.py:957]: why do we need padding here and `padding_side="left"` ? We just forward a single sequence no?

File: src/transformers/generation/utils.py [hunk 9]
```diff
@@ -914,42 +914,121 @@ def _merge_criteria_processor_list(
...
+        ...     num_return_sequences=4,
+        ...     return_dict_in_generate=True,
+        ...     output_scores=True,
+        ... )
+        >>> transition_scores = model.compute_transition_scores(
+        ...     outputs.sequences, outputs.scores, outputs.beam_indices, normalize_logits=False
+        ... )
+        >>> # If you sum the generated tokens' scores and apply the length penalty, you'll get the sequence scores.
+        >>> # Tip: set `normalize_logits=True` to recompute the scores from the normalized logits.
+        >>> output_length = inputs.input_ids.shape[1] + np.sum(transition_scores.numpy() < 0, axis=1)
```
Comment (patrickvonplaten) [inline src/transformers/generation/utils.py:989]: very nice!

**Review by patrickvonplaten — APPROVED** [review]
Thanks a lot for working on this - great job at improving the API & docs here so much :heart: 

Think it's totally fine to give it a different name and very nice to make it a first class citizen :-) Think this function could also go into a doc explaining advanced `generate`.

Also it'd be super nice if you could post your improvements and your example maybe on this thread: https://discuss.huggingface.co/t/generation-probabilities-how-to-compute-probabilities-of-output-scores-for-gpt2/3175/22 

I didn't do a good job back then at keeping the community up to date with this.

File: src/transformers/generation/utils.py (hunk 8 above)
Comment (gante) [reply src/transformers/generation/utils.py:957]: haha there is no need, leftovers from a copy paste -- going to remove it

**Comment by gante** [comment]
@sgugger merging as the failing test is a known flaky test (`tests/models/auto/test_modeling_auto.py::AutoModelTest::test_from_pretrained_dynamic_model_distant`)

**Comment by guberti** [comment]
Thanks! This function is super helpful for my use case.

**Comment by Ki-Seki** [comment]
OMG, I'm just hacking on making something like this. Thrilled to discover this function. Super helpful!
